Arm/BitVec.lean

```diff
@@ -813,4 +813,84 @@ macro_rules
 
 -- #print test2
 
+/-! ### Equation Lemmas -/
+
+theorem gt_def {x y : BitVec w} : (x > y) = (x.toNat > y.toNat) := rfl
+theorem ge_def {x y : BitVec w} : (x ≥ y) = (x.toNat ≥ y.toNat) := rfl
+
+theorem neq_of_toNat_neq {x y : BitVec w} (hx : x.toNat ≠ y.toNat) : x ≠ y := by
+  intros h
+  simp [h] at hx
+
+/-! ### (Non) Overflow Rewriting Lemmas
+
+@bollu: To be upstreamed.
+-/
+
+private theorem Nat.sub_add_eq_add_sub_of_le_of_le {a b c : Nat} (hab : b ≤ a) (hbc : b ≤ c) :
+    a - b + c = a + (c - b) := by
+  omega
+
+/- TODO: upstream. -/
+theorem toNat_sub_eq_toNat_sub_toNat_of_le {x y : BitVec w} (h : y ≤ x) :
+    (x - y).toNat = x.toNat - y.toNat := by
+  simp only [toNat_sub]
+  rw [BitVec.le_def] at h
+  by_cases h' : x.toNat = y.toNat
+  · rw [h', Nat.sub_self]
+    rw [Nat.sub_add_cancel (by omega), Nat.mod_self]
+  · rw [Nat.sub_add_eq_add_sub_of_le_of_le (by omega) (by omega), Nat.add_mod,
+      Nat.mod_self, Nat.zero_add, Nat.mod_mod, Nat.mod_eq_of_lt (by omega)]
+
+theorem neq_of_lt {x y : BitVec w₁} (h : x < y) : x ≠ y := by
+  rintro rfl
+  simp [BitVec.lt_def] at h
+
+theorem neq_of_gt {x y : BitVec w₁} (h : x > y) : x ≠ y := by
+  rintro rfl
+  simp [BitVec.lt_def] at h
+
+
+/-- adding bitvectors when there is no overflow. -/
+theorem toNat_add_eq_toNat_add_toNat {x y : BitVec w} (h : x.toNat + y.toNat < 2^w) :
+    (x + y).toNat = x.toNat + y.toNat := by
+  rw [BitVec.toNat_add, Nat.mod_eq_of_lt h]
+
+/-! ### Least Significant Byte -/
+
+/-- Definition to extract the `n`th least significant *Byte* from a bitvector. -/
+def extractLsByte (val : BitVec w₁) (base : Nat) : BitVec 8 :=
+  val.extractLsb ((base + 1)*8 - 1) (base * 8) |>.cast (by omega)
+
+theorem extractLsByte_def (val : BitVec w₁) (n : Nat) :
+    val.extractLsByte n = (val.extractLsb ((n + 1)*8 - 1) (n * 8) |>.cast (by omega)) := rfl
+
+@[simp]
+theorem getLsb_extractLsByte (val : BitVec w₁) :
+    ((BitVec.extractLsByte val n).getLsb i) =
+    (decide (i ≤ 7) && val.getLsb (n * 8 + i)) := by
+  simp [extractLsByte]
+  rw [Nat.succ_mul]
+  simp only [Nat.add_one_sub_one]
+  simp [Nat.add_sub_cancel_left]
+
+
+/-! ### Least Significant Byte range -/
+
+/-- grab 'n' least significant bytes, starting from index 'base'.
+@bollu: it's not clear if the definition for n=0 is desirable.
+-/
+def extractLsBytes (val : BitVec w) (base : Nat) (n : Nat) : BitVec (n * 8) :=
+  match h : n with
+  | 0 => 0#0
+  | x + 1 => val.extractLsb (base * 8 + n * 8 - 1) (base * 8) |>.cast (by omega)
+
+@[simp]
+theorem getLsb_extractLsBytes (val : BitVec w) (base : Nat) (n : Nat) (i : Nat) :
+    (BitVec.extractLsBytes val base n).getLsb i =
+      (decide (0 < n) && (decide (i ≤ base * 8 + (n) * 8 - 1 - base * 8) && val.getLsb (base * 8 + i))) := by
```

Comment:
Quibble: let's try to wrap long lines at column 80.

Reply:
[Mathlib allows column length 100](https://github.com/leanprover-community/mathlib4/blob/055d766c68bd10e70d7adf8fe87f44add3d806fa/Mathlib/Tactic/Linter/Lint.lean#L259C9-L260). Happy to switch to column length 80 if desired.

Reply:
Yes, please. Thanks!